### PR TITLE
Do not convert blkio weight value using blkio->io conversion scheme

### PR DESCRIPTION
**Pull request by dqminh**
bfq weight controller (i.e. io.bfq.weight if present) is still using the
same bfq weight scheme (i.e 1->1000, see [1].) Unfortunaly the
documentation for this was wrong, and only fixed recently [2].

Therefore, if we map blkio weight to io.bfq.weight, there's no need to
do any conversion.

[1] https://github.com/torvalds/linux/blob/master/Documentation/block/bfq-iosched.rst
[2] https://github.com/torvalds/linux/commit/65752aef0a407e1ef17ec78a7fc31ba4e0b360f9

### CODE REVIEW

**Comment by dqminh** [comment]
This will also need to be fixed in crun and systemd among other things afair @giuseppe

**Comment by giuseppe** [comment]
> This will also need to be fixed in crun and systemd among other things afair @giuseppe

thanks for letting me know!  I've opened a PR for crun as well: https://github.com/containers/crun/pull/584

**Comment by giuseppe** [comment]
@keszybz FYI

**Comment by dqminh** [comment]
Updated to add a test so we can try to catch this in the future. I'm not sure if there's movement to merge io.weight and io.bfq.weight at the moment, at which point we need to fix the inconsistency between 2 interface.

**Review by kolyshkin — APPROVED** [review]
LGTM

**Comment by kolyshkin** [comment]
CI failure on Fedora is a glitch (of download.fedoraproject.org) -- created https://github.com/opencontainers/runc/pull/2787 to address.

CI restared.

File: tests/integration/cgroups.bats
Comment (AkihiroSuda) [inline tests/integration/cgroups.bats:194]: Shouldn't require root?

File: tests/integration/cgroups.bats
Comment (cyphar) [inline tests/integration/cgroups.bats:199]: If you add a new container name you need to add it to the `teardown()` hook at the start -- `teardown_running_container`.

File: tests/integration/cgroups.bats
Comment (cyphar) [inline tests/integration/cgroups.bats:204]: This shouldn't be necessary? `runc` is a function which prints the output to the bats log (I'm also pretty sure a naked `echo` doesn't work which is why `runc` redirects to `stderr`):

```bash
# Wrapper for runc.
function runc() {
	run __runc "$@"

	# Some debug information to make life easier. bats will only print it if the
	# test failed, in which case the output is useful.
	echo "runc $@ (status=$status):" >&2
	echo "$output" >&2
}

# Raw wrapper for runc.
function __runc() {
	"$RUNC" ${RUNC_USE_SYSTEMD+--systemd-cgroup} --root "$ROOT" "$@"
}
```

**Review by cyphar — CHANGES_REQUESTED** [review]
Test needs to be cleaned up.

**Comment by keszybz** [comment]
@giuseppe thanks for the ping.

> This will also need to be fixed in crun and systemd among other things afair

Do I get this right that `io.bfq.weight` wants a value that is in the range 1..1000, while `io.weight` wants a value in the range 1..10000? systemd treats the cgroupv2 range as the input value, so we'd need to add a conversion `io.bfq.weight = (io.weight + 9) / 10`.

**Comment by keszybz** [comment]
^ @paolo-github

**Comment by keszybz** [comment]
https://github.com/systemd/systemd/pull/18467

**Comment by dqminh** [comment]
@keszybz yes, i think that is correct.

Which then also raises more questions:

1. Do we want to set both interface ( like what systemd is doing ) ? In our case, since the spec is saying that the allowed range is 0-1000, and i assume that we don't want breaking changes, we still have to do the conversion to write to `io.weight`
2. I don't think runc's systemd v2 is setting IOWeight, is that right ? I tried to do a quick search and can't find any.

In general, `io.weight` iocost interface is also more likely to be available everywhere (is it ?, see https://github.com/containers/podman/issues/8737), so maybe doing 1) makes sense.

File: tests/integration/cgroups.bats
Comment (kolyshkin) [reply tests/integration/cgroups.bats:199]: I noticed it, too, but did not comment because I am fixing all this soon (after https://github.com/opencontainers/runc/pull/2757 is merged).

**Review by kolyshkin — COMMENTED** [review]
Indeed, the `io.bfq.weight` range is 1-1000, but `io.weight` range is 1-10000. runc is not currently setting `io.weight`.

**Comment by paolo-github** [comment]
If useful for you, I'll be glad to change the range for BFQ too. The current BFQ range is due just to legacy.

**Comment by keszybz** [comment]
> If useful for you, I'll be glad to change the range for BFQ too. The current BFQ range is due just to legacy.

That'd make an even bigger mess ;( Please just make bfq read `io.weight`...

**Comment by paolo-github** [comment]
Ok, I don't want to add noise

Yes, making bfq read io.weight is the sensible thing. We fought for that, for months. Not only we pushed, but we also proposed three or four working solutions, following the always varying requests. At no avail, as I said in the other thread.

I don't think I can do more than that, sorry.

**Comment by kolyshkin** [comment]
nit: a typo in commit message: `Unfortunaly -> Unfortunately`.

The only non-addressed review comment is https://github.com/opencontainers/runc/pull/2786#discussion_r569835244

**Review by AkihiroSuda — APPROVED** [review]
Nits can be fixed later. LGTM.

**Comment by kolyshkin** [comment]
Took a liberty to rebase, fix a typo in commit message, and remove root requirement from the integration test case. Let's see if it passes CI.

**Review by kolyshkin — APPROVED** [review]
LGTM

**Comment by kolyshkin** [comment]
@AkihiroSuda PTAL

**Comment by kolyshkin** [comment]
The test case is not working under rootless on Fedora:
```
not ok 10 runc run (blkio weight)
# (in test file /vagrant/tests/integration/cgroups.bats, line 203)
#   `[ "$status" -eq 0 ]' failed
# runc spec --rootless (status=0):
# 
# runc run -d --console-socket /tmp/console.sock test_cgroups_unified (status=0):
# 
# runc exec test_cgroups_unified sh -c cat /sys/fs/cgroup/io.bfq.weight (status=1):
# cat: can't open '/sys/fs/cgroup/io.bfq.weight': No such file or directory
```

Apparently the problem is, rootless containers somehow do no honor cgroupns. Instead of "just this container" we see everything in container's view of /sys/fs/cgroup.

Looking...

**Comment by kolyshkin** [comment]
Normal container's cgroup2 mount:
> [kir@kir-rhat runc]$ cat /proc/2474364/mountinfo  | grep cgroup
> 1956 1955 0:27 /user.slice/user-1000.slice/user@1000.service/x453 /sys/fs/cgroup ro,nosuid,nodev,noexec,relatime - cgroup2 cgroup rw,seclabel

rootless container cgroup2 mount:
> [kir@kir-rhat runc]$ cat /proc/2475909/mountinfo  | grep cgroup
> 30 23 0:27 / /sys/fs/cgroup rw,nosuid,nodev,noexec,relatime shared:4 - cgroup2 cgroup2 rw,seclabel

Ah, so it is caused by https://github.com/opencontainers/runc/issues/2158. Hmmm.

**Comment by dqminh** [comment]
@kolyshkin sorry for the delay. yes i indeed hit this along the removal of root for the test so it wasn't done, and i forgot to put a comment on that.

**Comment by kolyshkin** [comment]
@dqminh your patch is fine (note that I pushed into your branch though), the rootless cgroup mount is a separate issue I'm going to address.

File: tests/integration/cgroups.bats
Comment (kolyshkin) [reply tests/integration/cgroups.bats:194]: @AkihiroSuda once https://github.com/opencontainers/runc/pull/2818 is merged we will be able to enable it (rm root, add `[[ "$ROOTLESS" -ne 0 ]] && requires rootless_cgroup`)

Comment (kolyshkin) [reply tests/integration/cgroups.bats:194]: ...because currently `runc exec test_cgroups_unified sh -c 'cat /sys/fs/cgroup/io.bfq.weight'` is not working

**Review by kolyshkin — APPROVED** [review]
LGTM

**Comment by kolyshkin** [comment]
@AkihiroSuda PTAL

**Review by AkihiroSuda — APPROVED** [review]

**Comment by kolyshkin** [comment]
@cyphar PTAL (you requested changes and thus merging is blocked until you approve). Seems your review comments are now addressed.

**Review by cyphar — APPROVED** [review]
LGTM. As an aside, feel free to dismiss my old reviews if they are clearly no longer applicable. (I wish GitHub did this already, because it treats approvals as something that can go stale but "request changes" reviews aren't treated as stale.)

**Pull request closed without merging**
